CreeDictionary/search_quality/analyze_results.py

```diff
@@ -0,0 +1,191 @@
+            [
+                r["matched_cree"],
+                r["lemma_wordform"]["text"],
+                r["lemma_wordform"]["inflectional_category"],
+                ",".join(r["raw_suffix_tags"]),
+            ]
+        )
+        if normalized in seen:
+            dupllcate_results += 1
+
+            seen[normalized]["is_duplicate_of"] = f"first {normalized}"
+            r["is_duplicate_of"] = normalized
+
+            if normalized not in unique_duplicates:
+                unique_duplicates.add(normalized)
+
+        else:
+            seen[normalized] = r
+
+    return {"total": dupllcate_results, "unique": len(unique_duplicates)}
+
+
+def find_rank(word: str, results: list[SerializedSearchResult]):
+    if not word.strip():
+        return ""  # n/a
+
+    index = 1
+    for r in results:
+        # We also have the definition here, could include it
+        if word == r["matched_cree"]:
+            return index
+        index += 1 if r["is_lemma"] else 2
+    return "☹"
+
+
+def load_results_file(results_file: PathLike) -> SampleSearchResultsJson:
+    with gzip.open(results_file, "rb") as f:
+        search_results: SampleSearchResultsJson = load(f)
+        return search_results
+
+
+def analyze(results_file, sample_definition: SampleDefinition = None):
+    """
+
+    If sample_definition is None, the default will be used.
+    """
+
+    if sample_definition is None:
+        sample_definition = load_sample_definition(DEFAULT_SAMPLE_FILE)
+
+    search_results = load_results_file(results_file)
+
+    analysis: Analysis = {}
+
+    for row in sample_definition:
+        keyword = row["English"]
+        keyword_search_results = search_results.get(keyword, None)
+
+        if keyword_search_results is None:
+            analysis[keyword] = None
+
+        else:
+            result_list = keyword_search_results["results"]
+
+            dupe_info = count_and_annotate_dupes(result_list)
+
+            query_analysis: SingleQueryAnalysis = {
+                "time_taken_seconds": keyword_search_results["time_taken_seconds"],
+                "result_count": count_results(result_list),
+                "total_duplicate_count": dupe_info["total"],
+                "unique_duplicate_count": dupe_info["unique"],
+            }
+
+            # How many top expected results does the sample specify for this
+            # query?
+            specified_result_count = 0
+            result_found_count = 0
+            threeten_good_result_count = 0
+
+            for i in [i + 1 for i in range(3)]:
+                # mypy does not like dynamic keys
+                word = cast(dict, row)[f"Nêhiyawêwin_{i}"]
+
+                if word:
+                    cast(dict, query_analysis)[f"word_{i}"] = word
+                    rank = find_rank(word, result_list)
+                    cast(dict, query_analysis)[f"word_{i}_rank"] = rank
+
+                    specified_result_count += 1
+                    if type(rank) == int:
```

Comment:
I think this is the more common way to check the type:

```suggestion
                    if isinstance(rank, int):
```

Reply:
Yup, thanks